Make Object ID and Group ID delta encoded in Fetch responses (#1586)
moq-wg/moq-transport: draft-ietf-moq-transport.md

conditions that constitute a malformed track when detected by a receiver
include:

> 1.  In a FETCH response, an Object with a particular Subgroup ID is received, but its

Author: Q: Is this really Fetch specific?  Can't this happen with Subscribe as well?

Bitmask | Condition if set | Condition if not set (0)
--------|------------------|---------------------

> 0x04 | Object ID Delta is present | Object ID is the prior Object's ID plus one

Reviewer: I think I meant 0x08 then?  But the current text looks right for Ascending/Descending.  The change is still annoying at least in my implementation.

0x10 | Priority field is present | Priority is the prior Object's Priority
0x20 | Properties field is present | Properties field is not present
0x40 | Datagram: ignore the two least significant bits | Decode the Subgroup ID as indicated by the two least significant bits

The first Object MUST include a Group ID Delta and Object ID Delta, and
these values are the absolute Group ID and Object ID. If the first Object in
the FETCH response uses a flag that references fields in the prior Object,
the Subscriber MUST close the session with a `PROTOCOL_VIOLATION`.

If the Group ID Delta field is present on an Object other than the first, the
Group ID is computed from the Group ID Delta and the prior Object's Group ID.
If the Group Order is Ascending, the Group ID is the prior Object's Group ID
plus the Group ID Delta + 1.  If the Group Order is Descending, the Group ID is
the prior Object's Group ID minus the (Group ID Delta + 1). If the computed
Group ID would be less than 0 or greater than 2^64-1, the Subscriber MUST
close the Session with error 'PROTOCOL_VIOLATION'.

If the Object ID Delta field is present on an Object other than the first in the

> Group, the Object ID is the prior Object's ID plus the Object ID Delta.  When the

Reviewer: First object in the group is ambiguous - eg if the group starts at object 0 but the fetch start is object 10.  I think this paragraph could be tighter in explaining the cases.

Group ID Delta Present -> Object ID Delta = [delta from 0]
Group ID Delta Not Present -> Object ID Delta [delta from last object ID in the group]

Author: PTAL a the suggestions below to see if they're correct?

0x10 | Priority field is present | Priority is the prior Object's Priority
0x20 | Properties field is present | Properties field is not present
0x40 | Datagram: ignore the two least significant bits | Decode the Subgroup ID as indicated by the two least significant bits

The first Object MUST include a Group ID Delta and Object ID Delta, and
these values are the absolute Group ID and Object ID. If the first Object in
the FETCH response uses a flag that references fields in the prior Object,
the Subscriber MUST close the session with a `PROTOCOL_VIOLATION`.

If the Group ID Delta field is present on an Object other than the first, the
Group ID is computed from the Group ID Delta and the prior Object's Group ID.
If the Group Order is Ascending, the Group ID is the prior Object's Group ID
plus the Group ID Delta + 1.  If the Group Order is Descending, the Group ID is
the prior Object's Group ID minus the (Group ID Delta + 1). If the computed
Group ID would be less than 0 or greater than 2^64-1, the Subscriber MUST
close the Session with error 'PROTOCOL_VIOLATION'.

If the Object ID Delta field is present on an Object other than the first in the

> Group, the Object ID is the prior Object's ID plus the Object ID Delta.  When the

Author: ```suggestion
If there is a prior Object in the Group and the Object ID Delta field is present,
the Object ID is the prior Object's ID plus the Object ID Delta.  When the
```

Group ID Delta field is present and therefore the Group ID changes, the Object ID

> is the absolute value of Object ID Delta if present, and 0 otherwise. If the computed

Author: ```suggestion
Group ID Delta field is present and therefore the Group ID changes, the Object ID
is the absolute value of Object ID Delta. If the computed
```

Object ID would be greater than 2^64-1, the Subscriber MUST close the Session

> with error 'PROTOCOL_VIOLATION'.

Reviewer: ```suggestion
When the Group ID Delta field is present, the Object ID is the value of Object ID Delta if
present. When the Group ID Delta field is not present, the Object ID is the prior Object's ID
plus the Object ID Delta if present. If Object ID Delta is not present, the Object ID is the
prior Object's ID plus one, regardless of which group it belongs to. If the computed Object ID
would be greater than 2^64-1, the Subscriber MUST close the Session with error
'PROTOCOL_VIOLATION'.
```
